### PR TITLE
Implement Mii Selector applet

### PR DESCRIPTION
**Pull request by Techie-Pi**
I've implemented the Mii Selector applet, I have yet to add documentation, but I would like to see your opinions on the API.
Specially around the creation and usage of the enums ``MiiConfigIndex`` and ``MiiType``.

### CODE REVIEW

**Comment by Techie-Pi** [comment]
Also, out of curiosity, why is [``MiiData`` an opaque type](https://github.com/rust3ds/ctru-rs/blob/9cf5e4de94e8d3711fd757e3464dd1880d91b92f/ctru-sys/bindgen.sh#L17)?
Nevermind, just saw the [issue](https://github.com/rust-lang/rust-bindgen/issues/1556), that's actually sad tbh

**Comment by Techie-Pi** [comment]
I've been researching, and the representation of the MiiData given by the 3DS doesn't change at all and [is documented](https://www.3dbrew.org/wiki/Mii#Mii_format). The purpose of libctru's ``MiiData`` is to _imitate_ the representation so that the struct is generated directly from the 3DS' MiiData.

Instead of that, we could parse the raw data to a Rust-compatible struct. I'll do some testing

**Comment by Techie-Pi** [comment]
Looks like it is possible, and quite simple tbh, maybe I could also add this to the PR.
Edit: Done... that was easier than what I had thought... I'll clean-up the code, fix a few things, and I'll incorporate the code into the PR

File: ctru-rs/src/mii.rs
Comment (Meziu) [inline ctru-rs/src/mii.rs]: Remove the `Console` prefix from all variants.

Comment (Techie-Pi) [reply ctru-rs/src/mii.rs]: But Rust doesn't allow numerals at the start of a member (3DS)

File: ctru-rs/src/mii.rs
Comment (Meziu) [inline ctru-rs/src/mii.rs]: Use the `ctru-sys` type rather than `[u8;92]`. Do this for all the cases in which it appears.

File: ctru-rs/src/mii.rs
Comment (Meziu) [reply ctru-rs/src/mii.rs]: True, hadn't noticed... Still, does "WiiUSwitch" actually exist as an option? I don't see it mentioned in `libctru`'s docs.

File: ctru-rs/src/mii.rs
Comment (Meziu) [inline ctru-rs/src/mii.rs]: `None` would be less repetitive.

File: ctru-rs/src/mii.rs
Comment (Techie-Pi) [reply ctru-rs/src/mii.rs]: The 3dbrew wiki says so.
[Ref](https://www.3dbrew.org/wiki/Mii#Mii_format) offset 0x3, bit 4-6

File: ctru-rs/src/mii.rs
Comment (Techie-Pi) [reply ctru-rs/src/mii.rs]: The problem is that [there is no ``ctru-sys`` type for ``[u8; 92]``](https://github.com/rust3ds/ctru-rs/blob/9cf5e4de94e8d3711fd757e3464dd1880d91b92f/ctru-sys/src/bindings.rs#L17597), but I could make a type alias

**Comment by Meziu** [comment]
My system's main mii is grayed out when selecting and thus can't be used. Is there an issue with the configuration?

File: ctru-rs/examples/mii-selector.rs
Comment (Meziu) [inline ctru-rs/examples/mii-selector.rs:14]: Why blacklist the user's mii?

File: ctru-rs/examples/mii-selector.rs
Comment (Meziu) [inline ctru-rs/examples/mii-selector.rs]: This message and the "author" message are both on row 5. In this order the "Moles" message wipes the "Author" one.

File: ctru-rs/examples/mii-selector.rs
Comment (Techie-Pi) [reply ctru-rs/examples/mii-selector.rs:14]: That's just an example, [the 3ds-examples' example also does this](https://github.com/devkitPro/3ds-examples/blob/cde9dff910f1115de79752cf55c609a9616cdc94/mii_selector/source/main.c#L54)

File: ctru-rs/src/applets/mii_selector.rs
Comment (Meziu) [inline ctru-rs/src/applets/mii_selector.rs]: We shouldn't hold this pointer. It has no use after being consumed by our wrapper.

File: ctru-rs/src/applets/mii_selector.rs
Comment (Meziu) [inline ctru-rs/src/applets/mii_selector.rs]: These functions are duplicate features already present (name and author can be obtained by the `MiiData`).

Furthermore, the checksum (while useful in a C environment) is a case we just shouldn't expect the user to check. The case of a faulty checksum means that our `MiiSelectorReturn` wrapped value is invalid, which is one of the safety measures we should prevent *by default*.

The checksum should be checked before returning the mii data to the user, and if the check fails, it should return an error. This check should be handled in the `MiiSelector::launch` method.

File: ctru-rs/src/applets/mii_selector.rs
Comment (Meziu) [inline ctru-rs/src/applets/mii_selector.rs:67]: All of the values in this struct should be `read-only`, since this is a *result* type for our operations. Rather than being `pub`, they should each have a corresponding `get` method.

Also, as stated in another comment, the presence of `checksum` here is to be discussed (as its validity would be already checked), thought I'd like to hear reasons why it may actually be useful, since I've never played around with this part of `libctru`.

File: ctru-rs/src/mii.rs
Comment (Meziu) [reply ctru-rs/src/mii.rs]: You should just use `ctru_sys::MiiData` for the conversion and take the inner buffer *here* rather than in `MiiSelectorResult`.

File: ctru-rs/src/mii.rs
Comment (Meziu) [inline ctru-rs/src/mii.rs]: I think you should remove all the `mii_` prefixes in the struct's members. Most of the time they will be called directly from the `mii_data.[member]` signature.

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [reply ctru-rs/src/mii.rs]: You can use `N3DS` instead

File: ctru-rs/examples/mii-selector.rs
Comment (AzureMarker) [inline ctru-rs/examples/mii-selector.rs]: I don't think the line positioning code is needed. The console automatically handles newlines.

Edit: Running this on my 3DS, the second line overwrites the first line until you remove the line positioning code.

**Review by Meziu — CHANGES_REQUESTED** [review]

File: ctru-rs/src/applets/mii_selector.rs
Comment (AzureMarker) [inline ctru-rs/src/applets/mii_selector.rs]: nit: based on the comment the `expect` error message could be more accurate/useful.

File: ctru-rs/src/applets/mii_selector.rs
Comment (AzureMarker) [inline ctru-rs/src/applets/mii_selector.rs]: nit: Could reduce duplication (this applies to multiple methods):
```suggestion
        let index = match mii_index {
            MiiConfigIndex::Index(i) => i,
            MiiConfigIndex::All => ctru_sys::MIISELECTOR_GUESTMII_SLOTS,
        };
        unsafe { ctru_sys::miiSelectorWhitelistGuestMii(self.config.as_mut(), index) };
```

Even better, move the `match` conversion to a method on `MiiConfigIndex`.

File: ctru-rs/src/applets/mii_selector.rs
Comment (AzureMarker) [inline ctru-rs/src/applets/mii_selector.rs]: Why include this in the struct if it's private? This struct will only be used by users, who can't access this field.

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs]: nit: This could be simpler:
```suggestion
fn utf16_byte_pairs_to_string(data: &[u8]) -> String {
    let raw_utf16_composed = data
        .chunks_exact(2)
        .map(|chunk| u16::from_le_bytes([chunk[0], chunk[1]]))
        .collect::<Vec<u16>>();

    String::from_utf16_lossy(raw_utf16_composed.as_slice())
}
```

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs]: nit: This could be simpler:
```suggestion
fn get_and_concat_vec_bit(data: &[u8], get_values: &[usize]) -> Vec<bool> {
    get_values
        .iter()
        .flat_map(|v| vec_bit(data[*v]))
        .collect()
}
```

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs]: This can just be `let raw_utf16_name = &raw_mii_data[0x1A..0x2D];` (`as_slice` isn't needed)

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs]: nit: it's more idiomatic to do `&raw_mii_data` instead of `raw_mii_data.as_slice()`

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs]:
```suggestion
            is_copying_allowed: raw_options[0],
            is_profanity_flag_enabled: raw_options[1],
```

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs]: Do this refactoring for the other similar instances below as well.
```suggestion
                match (raw_options[3], raw_options[2]) {
                    (false, false) => RegionLock::None,
                    (false, true) => RegionLock::Japan,
                    (true, false) => RegionLock::USA,
                    (true, true) => RegionLock::Europe,
                }
```

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs]:
```suggestion
                if raw_details[0] {
                    MiiSex::Female
                } else {
                    MiiSex::Male
                }
```

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs]: This function's functionality is still unclear to me, maybe a better name and more descriptive docs would help?

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs]: nit: This could be simplified:
```suggestion
fn vec_bit_to_u8(data: [bool; 8]) -> u8 {
    data.into_iter().fold(0, |result, bit| {
        (result << 1) ^ u8::from(bit)
    })
}
```

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs:8]: It makes sense to derive `Copy` for most of these structs/enums.

File: ctru-rs/src/applets/mii_selector.rs
Comment (AzureMarker) [reply ctru-rs/src/applets/mii_selector.rs:67]: I think it's fine to return the whole struct and let the user mess with it. It's not the library's problem after it gets returned to the user. They can muck around with it all they want, we don't need it back :).

File: ctru-rs/src/applets/mii_selector.rs
Comment (Techie-Pi) [reply ctru-rs/src/applets/mii_selector.rs]: Oh yeah, that's something that I left from testing, I'll remove it
Edit: Actually, it was used. But I've removed it

File: ctru-rs/src/mii.rs
Comment (Techie-Pi) [reply ctru-rs/src/mii.rs]: Ohhh that's actually great! I hadn't thought of doing that

File: ctru-rs/src/applets/mii_selector.rs
Comment (Techie-Pi) [reply ctru-rs/src/applets/mii_selector.rs]: > Even better, move the match conversion to a method on MiiConfigIndex.

The problem with that is that, depending on the context, ``MiiConfigIndex::All`` returns ``MIISELECTOR_GUESTMII_SLOTS`` or ``MIISELECTOR_USERMII_SLOTS``

File: ctru-rs/src/applets/mii_selector.rs
Comment (Techie-Pi) [reply ctru-rs/src/applets/mii_selector.rs]: Is it better now?

File: ctru-rs/src/mii.rs
Comment (Techie-Pi) [reply ctru-rs/src/mii.rs:450]: Maybe I need help writing the docs for this function, so I'll write what's the purpose of this functions here so you'll be able to give some suggestions.
The 3DS is a LE system, but for humans it's _easier_ to read and understand BE bits. So, this function is given a BE sequence of bits (e.g. ``[true, true, false]``) that gets: (i) filled with _``false``s_ to fill the remaining space until 8 bits (``u8``) are filled (e.g. ``[false, false, false, false, false, true, true, false]``), and (ii) the bits returned are transformed to be LE (e.g. ``[false, true, true, false, false, false, false, false]`` so the 3DS can understand them.

``partial_vec`` -> A Vec\<bool> (_in reality it is a slice_) that doesn't fill the 8 bits of an ``u8``
``with_reverse`` -> Where the values being returned are LE (""_reversed_"")

Maybe something like ``partial_bits_to_le_u8`` is better?

File: ctru-rs/src/mii.rs
Comment (Techie-Pi) [reply ctru-rs/src/mii.rs]: > You can use N3DS instead

Just to clarify (_sorry it is late here_), ``N3DS`` instead of ``WiiUSwitch`` right?

File: ctru-rs/src/applets/mii_selector.rs
Comment (Techie-Pi) [reply ctru-rs/src/applets/mii_selector.rs:67]: I agree with AzureMaker, and the ``checksum`` thing has been solved

File: ctru-rs/src/mii.rs
Comment (Meziu) [reply ctru-rs/src/mii.rs]: No, write `N3DS` instead of `Console3DS`. At that point you can remove all the `Console` prefixes.

Comment (Techie-Pi) [reply ctru-rs/src/mii.rs]: Alright, that makes sense, idk why I thought that it was meant to replace ``WiiUSwitch``.
About replacing ``Console3DS`` with ``N3DS``, wouldn't people confuse ``N3DS`` with New 3DS?

File: ctru-rs/src/mii.rs
Comment (Techie-Pi) [reply ctru-rs/src/mii.rs]: Alright, that function was _actually_ misused. The purpose of the function was writing BE bits, filling them until the length was 8 bits and making them LE.

This function was used this way before, but now the data is directly from the 3DS and this isn't needed + it caused some _big big_ bugs. I've fixed it and documented it

**Comment by Techie-Pi** [comment]
(For some reason I can't request multiple reviews at the same time /shrug)

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [reply ctru-rs/src/mii.rs]: We can use `New3DS` or `NewN3DS` for that. `N3DS` could mean "Nintendo 3DS"

Comment (AzureMarker) [reply ctru-rs/src/mii.rs]: Or use `O3DS` for "Original 3DS", but I don't think that's used often.

I named one of my repos [`n3ds-controller`](https://github.com/AzureMarker/n3ds-controller/) because of this issue.

File: ctru-rs/src/applets/mii_selector.rs
Comment (AzureMarker) [reply ctru-rs/src/applets/mii_selector.rs]: In that case just doing the refactor in each method would be good.

File: ctru-rs/src/applets/mii_selector.rs
Comment (AzureMarker) [inline ctru-rs/src/applets/mii_selector.rs]: Is the checksum relevant now that the original data structure isn't returned?

File: ctru-rs/src/applets/mii_selector.rs
Comment (AzureMarker) [inline ctru-rs/src/applets/mii_selector.rs]: nit: this should be `From<ctru_sys::MiiSelectorReturn>`. The `Box` part isn't needed/utilized.

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs]: Are you sure `is_sharing_enabled` uses `raw_face_style[1]`? `shape` also uses that value.

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs]: I see some missed bits here. `1..4` is `1, 2, 3`, and `5..9` is `5, 6, 7, 8` (4 is missing).
This shows up a lot, is it expected?

Comment (Techie-Pi) [reply ctru-rs/src/mii.rs]: Yeah this is expected, some bits are completely unused/padding according to 3dbrew

File: ctru-rs/src/applets/mii_selector.rs
Comment (Techie-Pi) [reply ctru-rs/src/applets/mii_selector.rs]: Oh yeah, I forgot to remove that

File: ctru-rs/src/mii.rs
Comment (Techie-Pi) [reply ctru-rs/src/mii.rs]: I think I'll use N3DS and clarify this in the documentation, seems like the best option to me

File: ctru-rs/src/applets/mii_selector.rs
Comment (Meziu) [inline ctru-rs/src/applets/mii_selector.rs:57]: `MiiSelectorReturn` should be `non_exhaustive` to prevent the user from creating it without using the service.

**Review by Meziu — CHANGES_REQUESTED** [review]

File: ctru-rs/src/mii.rs
Comment (Techie-Pi) [reply ctru-rs/src/mii.rs]: Hmm now that I see that maybe you're right here, I had misread what you said :p

Comment (Techie-Pi) [reply ctru-rs/src/mii.rs]: Turns out you were completely right! The values I were using for testing just didn't trigger any bug

**Comment by Techie-Pi** [comment]
I've squashed some commits (mostly docs and cargo fmt/clippy)

**Review by Meziu — APPROVED** [review]
Looks good to me!

File: ctru-rs/src/applets/mii_selector.rs
Comment (AzureMarker) [inline ctru-rs/src/applets/mii_selector.rs]: nit: This variable isn't needed, you can directly reference `ret.guest_mii_index`

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs:359]: Are there supposed to be gaps here?

File: ctru-rs/src/mii.rs
Comment (AzureMarker) [inline ctru-rs/src/mii.rs:450]: If these are leading zeros, why aren't they at the start of the list?

File: ctru-rs/src/mii.rs
Comment (Techie-Pi) [reply ctru-rs/src/mii.rs:359]: Yes, here it is

File: ctru-rs/src/mii.rs
Comment (Techie-Pi) [reply ctru-rs/src/mii.rs:450]: Because of the way the bits are ordered by the 3ds, (this time I tested thoroughly to make sure the values returned are correct, so this should always work :p, but I'll double check later)

Comment (Techie-Pi) [reply ctru-rs/src/mii.rs:450]: Yeah. Tested again and it works properly

**Review by ian-h-chamberlain — APPROVED** [review]
Sorry I meant to review this a while ago, but finally got a chance to take a look and it seems pretty cool! 

Some of `ctru::MiiData` might be worth documenting further, if we're going to maintain a totally separate struct vs the `ctru_sys::MiiData`, but I think this is good to start. There's lots of documentation stuff that needs to be done across the board so I don't think it should keep this from merging as-is.

**Comment by Techie-Pi** [comment]
After current concerns about this PR are resolved, I plan on adding (at least) some basic documentation (and maybe references to the 3dbrew wiki) before merging, though I think the best thing would be to merge this PR with basic documentation and add more detailed documentation in a follow-up PR (I'm not a native English speaker, so I may need help in the follow-up PR, which may take some time to make because of that :p)

**Review by AzureMarker — APPROVED** [review]

**Comment by Techie-Pi** [comment]
Is there anything that should be documented in this PR instead of the follow-up? Or is this ready to merge?

**Comment by Meziu** [comment]
> Is there anything that should be documented in this PR instead of the follow-up? Or is this ready to merge?

I believe it’s fine for now. PR anything you want to add.

**Pull request merged**
